### ISSUE TITLE
Support SpeciesNet Ensemble

### ISSUE THREAD

**Issue by geohacker** [tnc-ca-geo/animl-ml#130]
ANIML now runs SpeciesNet (#126) support for `all`, `classifier` and `detector` components. SpeciesNet has an [`ensemble` component that combines](https://github.com/google/cameratrapai?tab=readme-ov-file#ensemble-decision-making) detector and classifier, along with location that provides high confidence predictions and label rollup. This is presently not enabled in ANIML.

Ensemble component requires a different payload — [combining output from detector and classifier](https://github.com/google/cameratrapai?tab=readme-ov-file#full-ensemble):

```
{
    "predictions": [
        {
            "filepath": str  => Image filepath.
            "failures": list[str] (optional)  => List of internal components that failed during prediction (e.g. "CLASSIFIER", "DETECTOR", "GEOLOCATION"). If absent, the prediction was successful.
            "country": str (optional)  => 3-letter country code (ISO 3166-1 Alpha-3) for the location where the image was taken. It can be overwritten if the country from the request doesn't match the country of (latitude, longitude).
            "admin1_region": str (optional)  => First-level administrative division (in ISO 3166-2 format) within the country above. If not provided in the request, it can be computed from (latitude, longitude) when those coordinates are specified. Included in the response only for some countries that are used in geofencing (e.g. "USA").
            "latitude": float (optional)  => Latitude where the image was taken, included only if (latitude, longitude) were present in the request.
            "longitude": float (optional)  => Longitude where the image was taken, included only if (latitude, longitude) were present in the request.
            "classifications": {  => dict (optional)  => Top-5 classifications. Included only if "CLASSIFIER" if not part of the "failures" field.
                "classes": list[str]  => List of top-5 classes predicted by the classifier, matching the decreasing order of their scores below.
                "scores": list[float]  => List of scores corresponding to top-5 classes predicted by the classifier, in decreasing order.
                "target_classes": list[str] (optional)  => List of target classes, only present if target classes are passed as arguments.
                "target_logits": list[float] (optional)  => Raw confidence scores (logits) of the target classes, only present if target classes are passed as arguments.
            },
            "detections": [  => list (optional)  => List of detections with confidence scores > 0.01, in decreasing order of their scores. Included only if "DETECTOR" if not part of the "failures" field.
                {
                    "category": str  => Detection class "1" (= animal), "2" (= human) or "3" (= vehicle) from MegaDetector's raw output.
                    "label": str  => Detection class "animal", "human" or "vehicle", matching the "category" field above. Added for readability purposes.
                    "conf": float  => Confidence score of the current detection.
                    "bbox": list[float]  => Bounding box coordinates, in (xmin, ymin, width, height) format, of the current detection. Coordinates are normalized to the [0.0, 1.0] range, relative to the image dimensions.
                },
                ...  => A prediction can contain zero or multiple detections.
            ],
            "prediction": str (optional)  => Final prediction of the SpeciesNet ensemble. Included only if "CLASSIFIER" and "DETECTOR" are not part of the "failures" field.
            "prediction_score": float (optional)  => Final prediction score of the SpeciesNet ensemble. Included only if the "prediction" field above is included.
            "prediction_source": str (optional)  => Internal component that produced the final prediction. Used to collect information about which parts of the SpeciesNet ensemble fired. Included only if the "prediction" field above is included.
            "model_version": str  => A string representing the version of the model that produced the current prediction.
        },
        ...  => A response will contain one prediction for each instance in the request.
    ]
}
```

For this to work reliably, we should:
* Run speciesnet in detector mode first. Store this similar how md5 does it perhaps
* Run speciesnet in classifier mode. Store this.
* Create a new type of trigger to run ensemble after the above, passing the a combined payload
* Handle the response and render

I've not spent a lot of time thinking about different places we might wire this up but we can dig into it when it's a priority.

cc @nathanielrindlaub

**Comment by nathanielrindlaub** [comment]
Update after chatting with @geohacker: we decided to not provide support for a stand-alone `ensemble` component in which animl-api is calling MegaDetector and SpeciesNet in separate, discrete steps, storing their detections + classifications, then passing just the results back into the SpeciesNet endpoint in `ensemble` mode to combine the results and perform the heuristic decision making [steps](https://github.com/google/cameratrapai?tab=readme-ov-file#ensemble-decision-making). 

The challenges is that the `ensemble` decision making logic expects _full_ results from _both_ models, and because Animl gives users a lot of control over each individual model invocation (they can turn off/on classes, adjust confidence thresholds), it doesn't store the full results of each inference request (we just select and store the highest confidence prediction that's above the confidence threshold), and often doesn't run species classifiers at all if MegaDetector doesn't detect an "animal", calling the SpecicesNet `ensemble` component in isolation and with incomplete detection inputs would probably produce some strange results. 

Instead, we are going to move forward by just offering the `all` mode, and adding selects to the Automation Rule config UI so that users can select their country and admin regions if they are running SpeciesNet in `all` mode.

**Comment by nathanielrindlaub** [comment]
@geohacker, I took a look at those images that SpeciesNet was returning "bird" predictions for in `classifier` mode that I thought might be a product of the classifier predicting other, more specific classifications for but that were getting ignored because it "bird" was higher confidence, but that was not the case. In all three images, there were indeed more specific classes predicted (e.g. "heron", "egret"), but they were super low confidence and would not have passed through any sensible confidence threshold filter. Also in all three cases, the second highest confidence prediction was "domestic cattle" at 2-8% confidence. So in these cases having the label roll up enabled wouldn't have changed the predictions:

```
{
    "predictions": [
        {
            "filepath": "/tmp/temp_image.jpg",
            "classifications": {
                "classes": [
                    "b1352069-a39c-4a84-a949-60044271c0c1;aves;;;;;bird",
                    "aca65aaa-8c6d-4b69-94de-842b08b13bd6;mammalia;cetartiodactyla;bovidae;bos;taurus;domestic cattle",
                    "ce2e0e86-5580-4ab4-974a-4970753a140f;aves;pelecaniformes;ardeidae;bubulcus;ibis;cattle egret",
                    "96fe1a07-7ef1-4a2f-99e1-ec2c9a78b532;aves;pelecaniformes;ardeidae;ardea;herodias;great blue heron",
                    "bfa75aeb-3187-48fe-95b9-f171465cc984;aves;pelecaniformes;ardeidae;ardea;cocoi;cocoi heron"
                ],
                "scores": [
                    0.9176809787750244,
                    0.07529006898403168,
                    0.0017425555270165205,
                    0.0008183515165001154,
                    0.00038497219793498516
                ]
            }
        }
    ]
}
```

![Image](https://github.com/user-attachments/assets/b1629cd1-8e89-4bec-9902-e25cfe824c38)
![Image](https://github.com/user-attachments/assets/22c02aa5-af71-4977-8098-837164cb53ec)
![Image](https://github.com/user-attachments/assets/3f9daaf5-ec56-4031-955f-91b5fb355f0f)

One thing I did catch was that when we run SpeciesNet in `all` mode, we should be using the `response.predictions[0].prediction` value instead of `response.predictions[0].classifications`, b/c the `prediction` is what the ensemble ultimately has decided is the most likely class.

We should still run it through `_filterClassifierPredictions()` though and use `response.predictions[0].prediction_source` and `response.predictions[0].model_version` as the `filteredDet.mlModel` and `filteredDet.mlModelVersion` when we're [creating](https://github.com/tnc-ca-geo/animl-api/blob/90ec18f54ceeab2283b32748ed50b6772ec67861/src/ml/modelInterfaces.ts#L53-L54) the RequestLabel payloads.

On second thought, I think your idea of having two separate MLModel records: one for `speciesnet-classifier` and one for `speciesnet-all` might be a smart move; that way we don't need to infer which model (Speciesnet or MDv5) actually made the prediction from `response.predictions[0].prediction_source`. I actually don't think we care which model made the prediction that was chosen, but I would like to keep track of whether that prediction was made in `all` mode or just came from the `classifier`.

Example SpeciesNet `all` mode prediction output:
```
{
    "predictions": [
        {
            "filepath": "/tmp/temp_image.jpg",
            "classifications": {
                "classes": [
                    "b1352069-a39c-4a84-a949-60044271c0c1;aves;;;;;bird",
                    "aca65aaa-8c6d-4b69-94de-842b08b13bd6;mammalia;cetartiodactyla;bovidae;bos;taurus;domestic cattle",
                    "d372cda5-a8ca-4b7b-97ed-4e4fab9c9b4b;mammalia;cetartiodactyla;suidae;sus;scrofa;wild boar",
                    "96fe1a07-7ef1-4a2f-99e1-ec2c9a78b532;aves;pelecaniformes;ardeidae;ardea;herodias;great blue heron",
                    "627c919c-29bc-439b-acde-62b3cd31f8a8;mammalia;cetartiodactyla;cervidae;;;cervidae family"
                ],
                "scores": [
                    0.963701605796814,
                    0.030378442257642746,
                    0.0008287281962111592,
                    0.0005858519580215216,
                    0.0004827631055377424
                ]
            },
            "detections": [
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.8808122277259827,
                    "bbox": [
                        0.5625,
                        0.3671875,
                        0.15087890625,
                        0.21484375
                    ]
                },
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.03670072928071022,
                    "bbox": [
                        0.56298828125,
                        0.4381510615348816,
                        0.10302734375,
                        0.134765625
                    ]
                },
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.03294273093342781,
                    "bbox": [
                        0.64208984375,
                        0.3665364496409893,
                        0.0712890625,
                        0.0963541641831398
                    ]
                }
            ],
            "prediction": "b1352069-a39c-4a84-a949-60044271c0c1;aves;;;;;bird",
            "prediction_score": 0.963701605796814,
            "prediction_source": "classifier",
            "model_version": "4.0.1a"
        }
    ]
}
```

**Comment by nathanielrindlaub** [comment]
## Update
Reopening this because I've discovered that SpeciesNet `all` mode is designed to produce a single class prediction for entire images rather than individual objects/bounding boxes. Admittedly, I should have seen this coming as the limitation is described in detail [here](https://github.com/google/cameratrapai/issues/8). Currently, Animl only [creates](https://github.com/tnc-ca-geo/animl-api/blob/fe747c92f3cf208df1fc59ea3d8b5b9ce684571f/src/ml/modelInterfaces.ts#L331) a single Object from the highest confidence bounding box (`detections[0]`) in the returned SpeciesNet payload and [applies](https://github.com/tnc-ca-geo/animl-api/blob/fe747c92f3cf208df1fc59ea3d8b5b9ce684571f/src/ml/modelInterfaces.ts#L335) the class returned in the `prediction` to that Object. This results in multi-species and multi-object images like the following only receiving one Object and Label (i.e. it correctly added an Object/Label for one dog but missed the other dog and the human):

![Image](https://github.com/user-attachments/assets/50db0e29-fc7a-4409-8e8b-dd8503f2f0f8)

```javascript
// Response from SpeciesNet in `all` mode for image above:

{
    "predictions": [
        {
            "filepath": "/tmp/temp_image.jpg",
            "country": "USA",
            "admin1_region": "CA",
            "classifications": {
                "classes": [
                    "3d80f1d6-b1df-4966-9ff4-94053c7a902a;mammalia;carnivora;canidae;canis;familiaris;domestic dog",
                    "5109acb4-e503-4147-a175-a3c6aa71f1e3;mammalia;perissodactyla;equidae;equus;caballus;domestic horse",
                    "d3d5325f-b055-4925-b875-5cd28af54215;mammalia;carnivora;canidae;canis;;canis species",
                    "aca65aaa-8c6d-4b69-94de-842b08b13bd6;mammalia;cetartiodactyla;bovidae;bos;taurus;domestic cattle",
                    "f1856211-cfb7-4a5b-9158-c0f72fd09ee6;;;;;;blank"
                ],
                "scores": [
                    0.7925304174423218,
                    0.09833765029907227,
                    0.035588644444942474,
                    0.024415146559476852,
                    0.008221253752708435
                ]
            },
            "detections": [
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.9349833726882935,
                    "bbox": [
                        0.5006250236183405,
                        0.4100000075995922,
                        0.058125000447034836,
                        0.058888889849185944
                    ]
                },
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.9124313592910767,
                    "bbox": [
                        0.32874999195337296,
                        0.3777777664363384,
                        0.08124999701976776,
                        0.0811111107468605
                    ]
                },
                {
                    "category": "2",
                    "label": "human",
                    "conf": 0.8804845809936523,
                    "bbox": [
                        0.02437499910593033,
                        0.25777777284383774,
                        0.04062499850988388,
                        0.18111111223697662
                    ]
                },
                {
                    "category": "1",
                    "label": "animal",
                    "conf": 0.22760775685310364,
                    "bbox": [
                        0.6049999999813735,
                        0.33555554784834385,
                        0.008750000037252903,
                        0.03222222253680229
                    ]
                }
            ],
            "prediction": "3d80f1d6-b1df-4966-9ff4-94053c7a902a;mammalia;carnivora;canidae;canis;familiaris;domestic dog",
            "prediction_score": 0.7925304174423218,
            "prediction_source": "classifier",
            "model_version": "4.0.1a"
        }
    ]
}
```

## Possible solutions
Ordered loosely by level of effort.

1. **Abandon support for `all` mode and the `ensemble` decision making, and just offer SpeciesNet in `classifier` mode:** In other words, do nothing; users would just have to configure their Automation Rules to run MegaDetector and SpeciesNet `classifier` in two discrete steps. This might be totally fine, but we would lose the ability to offer the SpeciesNet geofencing and label-rollup heuristics (see [this issue](https://github.com/tnc-ca-geo/animl-frontend/issues/338)). **Note: this is currently what we've reverted to doing.**

2. **Continue to support `all` mode, but assume all objects in an image are the same species:** expose a detection confidence threshold to the user, and for each detection SpeciesNet returns in `all` mode, create Objects for all of them above that threshold and assume that all detections are the same class. I imagine this is what WI does. TBH, although Wildlife Insights/SpeciesNet maintainers chose their one-class-per-image approach intentionally because multi-species images are relatively rare and it is less expensive to only run the classifier on the highest-confidence detectons, in my experience they occur frequently enough to make me feel uncomfortable about assuming that every object in an image is the same class. In addition, Animl was architected around the concept of object-level annotations (mostly to improve the quality of the datasets/crops for model training, but also because it's a more granular and accurate representation of the real world), so it would be a deviation from the pattern we use for all other classifiers in which we request individual species classifications for all detections above a certain threshold. 

3. **Abandon `all` mode, and implement our own geofencing mechanism:** Some options for that are described in [this](https://github.com/tnc-ca-geo/animl-frontend/issues/338) issue.

4. **Abandon `all` mode, but include an optional `ensemble` step for `classifier` mode requests:** Dan posted some workflows/notebooks [here](https://github.com/google/cameratrapai/issues/8#issuecomment-2702623993) for supporting multi-species images in which he demonstrates getting SpeciesNet `classifier` predictions for multiple detections and then running `run_model.py` with an `--ensemble_only` flag on all of those classifications to perform the heuristics on each individual detection + classification pair. We could take inspiration from that approach and do something similar. Admittedly, I had [concerns](https://github.com/tnc-ca-geo/animl-ml/issues/130#issuecomment-2933357784) about exposing a stand-alone `ensemble` mode endpoint or option because it seemed daunting to ensure we'd have all the necessary inputs, but I'm now warming up to it a bit. How I'm now imagining this working: 
   - offer a checkbox in the Automation Rule form to optionally perform `ensemble` decision making on SpeciesNet `classifier` predictions and expose `country` and `admin1region` selectors
   - if it's checked, in addition to [passing](https://github.com/tnc-ca-geo/animl-api/blob/fe747c92f3cf208df1fc59ea3d8b5b9ce684571f/src/ml/modelInterfaces.ts#L295-L305) the `bbox` into the SpeciesNet endpoint in the `animl-api` `modelInferface`, pass in more info about the `label`/detection including confidence and class, and then use that info on the FastAPI side to enrich the `detections_dict` [here](https://github.com/tnc-ca-geo/animl-ml/blob/e29aeb398c29546d8fab20d0f1b56040b522a145/api/speciesnet/serve.py#L111C1-L118C18). From there, immediately after calling [`model.classify()`](https://github.com/tnc-ca-geo/animl-ml/blob/e29aeb398c29546d8fab20d0f1b56040b522a145/api/speciesnet/serve.py#L120C1-L124C18), call `model.ensemble_from_past_runs()` as demonstrated in `run_model.py` [here](https://github.com/google/cameratrapai/blob/b1185e6a14fe76937abb7d87d1ae29296f8f0850/speciesnet/scripts/run_model.py#L433-L439). I think this should work ok, but I'm a little unsure if the juice is worth the squeeze. 
   One additional (maybe minor?) concern is that while we will be able to reasonably guarantee that we'll have crops/bboxes for any SpeciesNet `clasifier` run once we [implement](https://github.com/tnc-ca-geo/animl-frontend/issues/233) an `MLModel.expectsCrops` field, I worry about guaranteeing that the detection classes/`labelIds` we pass in will always match what's expected in the [`ensemble_prediction_combiner.py`](https://github.com/google/cameratrapai/blob/b1185e6a14fe76937abb7d87d1ae29296f8f0850/speciesnet/ensemble_prediction_combiner.py#L81). We could be certain the Ids could be [converted](https://github.com/google/cameratrapai/blob/main/speciesnet/constants.py) appropriately if MDv5a or MDv5b produced the detection, but if a user used a different object detection model in the future it would be tough to account for variations in classes and `labelIds` across detection models.

**Comment by geohacker** [comment]
@nathanielrindlaub Ah. Sorry, I should have caught that as well. I will mull about this.
